### PR TITLE
Add lamport clocks

### LINKED ISSUE #47: [bug] consensus breaks when two clients insert at the same position
Because we don't have any mechanism for finding consensus Xd.

### PR DESCRIPTION
**Pull request by katzper01**
Resolves #47

### CODE REVIEW

File: src/pages/main/collaboration/collaboration.ts [hunk 1]
```diff
@@ -23,17 +23,34 @@ export const useCollaboration = (projectId: string): Collaboration => {
...
+  const lamportClock = useRef<number>(0);
+  
+  useEffect(() => {
+    lamportClock.current = Math.max(lamportClock.current, state.initClock);
+  }, [state.initClock]);
 
   const nextCharacterIdRef = useRef<number>(0);
 
+  const handleIncomingClock = (incomingClock: number) => 
+    lamportClock.current = Math.max(lamportClock.current, incomingClock);
```
Comment (hubzie) [inline src/pages/main/collaboration/collaboration.ts:35]: Maybe I don't understand what there is on Wikipedia, but shouldn't there be
```suggestion
  const handleIncomingClock = (incomingClock: number) => 
    lamportClock.current = Math.max(lamportClock.current, incomingClock) + 1;
```
?

**Review by hubzie — APPROVED** [review]
Looks good but complicated

File: src/pages/main/collaboration/collaboration.ts (hunk 1 above)
Comment (katzper01) [reply src/pages/main/collaboration/collaboration.ts:35]: would work too, but every time I create new character i do

```
lamportClock++;
char.clock = lamportClock
```

So it also fine the way it is

Comment (hubzie) [reply src/pages/main/collaboration/collaboration.ts:35]: Okay, if you say so

**Pull request merged**